File: cmd/dlq/format.go
package main

import (
	"fmt"
	"os"
	"strings"
	"text/tabwriter"
	"time"
)

func printJobs(jobs []jobView) {
	if len(jobs) == 0 {
		fmt.Println("No jobs.")
		return
	}
	tw := tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
	fmt.Fprintln(tw, "ID\tSTATUS\tPROGRESS\tSPEED\tETA\tOUT\tNAME/URL")
	for _, j := range jobs {
		done := j.BytesDone
		if done == 0 && (j.Status == "completed" || j.Status == "decrypting" || j.Status == "decrypt_failed") && j.SizeBytes > 0 {
			done = j.SizeBytes
		}
		progress := formatProgress(done, j.SizeBytes)
		name := displayName(j)
		speed := formatSpeed(j)
		eta := formatETA(j)
		fmt.Fprintf(tw, "%d\t%s\t%s\t%s\t%s\t%s\t%s\n", j.ID, j.Status, progress, speed, eta, j.OutDir, name)
		if j.ErrorCode != "" {
			fmt.Fprintf(tw, " \t \t \t \t \t \t  error: %s (%s)\n", j.ErrorCode, j.Error)
		}
	}
	_ = tw.Flush()
}

func shortURL(u string) string {
	if len(u) > 64 {
		return u[:61] + "..."
	}
	return u
}

func displayName(j jobView) string {
	name := j.Filename
	if name == "" {
		name = j.Name
	}
	if name == "" {
		name = shortURL(j.URL)
	}
	return name
}

func filePath(j jobView) string {
	name := j.Filename
	if name == "" {
		name = j.Name
	}
	if name == "" {
		return shortURL(j.URL)
	}
	return strings.TrimRight(j.OutDir, "/") + "/" + name
}

func formatProgress(done, total int64) string {
	if total <= 0 {
		return humanBytes(done)
	}
	pct := float64(done) / float64(total) * 100
	return fmt.Sprintf("%s / %s (%.1f%%)", humanBytes(done), humanBytes(total), pct)
}

func hasActiveJobs(jobs []jobView) bool {
	for _, j := range jobs {
		switch j.Status {
		case "queued", "resolving", "downloading", "paused", "decrypting":
			return true
		}
	}
	return false
}

func humanBytes(n int64) string {
	units := []string{"B", "KiB", "MiB", "GiB", "TiB", "PiB"}
	val := float64(n)
	idx := 0
	for val >= 1024 && idx < len(units)-1 {
		val /= 1024
		idx++
	}
	return fmt.Sprintf("%.1f%s", val, units[idx])
}

func formatSpeed(j jobView) string {
	if j.Status != "downloading" || j.DownloadSpeed <= 0 {
		return "-"
	}
	return fmt.Sprintf("%s/s", humanBytes(j.DownloadSpeed))
}

func formatETA(j jobView) string {
	if j.Status != "downloading" || j.EtaSeconds <= 0 {
		return "-"
	}
	return humanDuration(time.Duration(j.EtaSeconds) * time.Second)
}

func humanDuration(d time.Duration) string {
	if d <= 0 {
		return "-"
	}
	seconds := int64(d.Seconds())
	h := seconds / 3600
	m := (seconds % 3600) / 60
	s := seconds % 60
	if h > 0 {
		return fmt.Sprintf("%dh%02dm", h, m)
	}
	if m > 0 {
		return fmt.Sprintf("%dm%02ds", m, s)
	}
	return fmt.Sprintf("%ds", s)
}

File: ui/src/lib/format.ts
import type { JobView } from './types';

const units = ['B', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB'];

export function humanBytes(n: number): string {
  if (!Number.isFinite(n) || n <= 0) {
    return '0.0B';
  }
  let val = n;
  let idx = 0;
  while (val >= 1024 && idx < units.length - 1) {
    val /= 1024;
    idx += 1;
  }
  return `${val.toFixed(1)}${units[idx]}`;
}

export function humanDuration(seconds: number): string {
  if (!Number.isFinite(seconds) || seconds <= 0) {
    return '-';
  }
  const s = Math.floor(seconds);
  const h = Math.floor(s / 3600);
  const m = Math.floor((s % 3600) / 60);
  const sec = s % 60;
  if (h > 0) {
    return `${h}h${String(m).padStart(2, '0')}m`;
  }
  if (m > 0) {
    return `${m}m${String(sec).padStart(2, '0')}s`;
  }
  return `${sec}s`;
}

export function filePath(job: JobView): string {
  const name = job.filename || job.name || shortURL(job.url);
  if (!name) {
    return shortURL(job.url);
  }
  return `${job.out_dir.replace(/\/$/, '')}/${name}`;
}

export function fileName(job: JobView): string {
  const name = job.filename || job.name || shortURL(job.url);
  if (!name) {
    return '-';
  }
  return name;
}

export function folderPath(job: JobView): string {
  const outDir = job.out_dir || '';
  if (!outDir || outDir === '/') {
    return outDir || '';
  }
  if (outDir.endsWith('/')) {
    return outDir;
  }
  return `${outDir}/`;
}

export function formatProgress(job: JobView): string {
  const total = job.size_bytes ?? 0;
  let done = job.bytes_done ?? 0;
  if (done === 0 && (job.status === 'completed' || job.status === 'decrypting' || job.status === 'decrypt_failed') && total > 0) {
    done = total;
  }
  if (total <= 0) {
    return humanBytes(done);
  }
  const pct = Math.min(100, (done / total) * 100);
  if (pct >= 100) {
    return humanBytes(total);
  }
  return `${humanBytes(done)} / ${humanBytes(total)} (${pct.toFixed(1)}%)`;
}

export function formatSpeed(job: JobView): string {
  if (job.status !== 'downloading' || !job.download_speed || job.download_speed <= 0) {
    return '-';
  }
  return `${humanBytes(job.download_speed)}/s`;
}

export function formatETA(job: JobView): string {
  if (job.status !== 'downloading' || !job.eta_seconds || job.eta_seconds <= 0) {
    return '-';
  }
  return humanDuration(job.eta_seconds);
}

export function shortURL(url: string): string {
  if (!url) return '';
  if (url.length <= 64) return url;
  return `${url.slice(0, 61)}...`;
}
